src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -825,24 +882,45 @@ mongoc_collection_estimated_document_count (
    }
 
    reply_ptr = reply ? reply : &reply_local;
-   bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
-
-   ret = _mongoc_client_command_with_opts (coll->client,
-                                           coll->db,
-                                           &cmd,
-                                           MONGOC_CMD_READ,
-                                           opts,
-                                           MONGOC_QUERY_NONE,
-                                           read_prefs,
-                                           coll->read_prefs,
-                                           coll->read_concern,
-                                           coll->write_concern,
-                                           reply_ptr,
-                                           error);
-
-   if (ret) {
-      if (bson_iter_init_find (&iter, reply_ptr, "n")) {
-         count = bson_iter_as_int64 (&iter);
+   if (server_stream->sd->max_wire_version < WIRE_VERSION_4_9) {
+      /* On < 4.9, use actual count command for estimatedDocumentCount */
+      bson_init (&cmd);
+      bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
+      ret = _mongoc_client_command_with_opts (coll->client,
+                                             coll->db,
+                                             &cmd,
+                                             MONGOC_CMD_READ,
+                                             opts,
+                                             MONGOC_QUERY_NONE,
+                                             read_prefs,
+                                             coll->read_prefs,
+                                             coll->read_concern,
+                                             coll->write_concern,
+                                             reply_ptr,
+                                             error);
+      if (ret) {
+         if (bson_iter_init_find (&iter, reply_ptr, "n")) {
+            count = bson_iter_as_int64 (&iter);
+         }
+      }
+   } else {
+      /* On >= 4.9, use aggregate with collStats for estimatedDocumentCount */
+      _make_aggregate_for_edc (coll, &cmd);
+      ret = mongoc_collection_read_command_with_opts (
+         coll, &cmd, read_prefs, opts, reply_ptr, error);
+
+      if (error && error->code == MONGOC_ERROR_COLLECTION_DOES_NOT_EXIST) {
+         /* Collection does not exist. From spec: return 0 but no err:
+          * https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#estimateddocumentcount
+          */
+         error = NULL;
+         count = 0;
+         GOTO (done);
```

Comment:
Although this swallows the error and returns 0, it does not touch the passed-in `reply` which still contains the NamespaceNotFound error. This is problematic, as users will not expect the reply to contain an error when the returned `error` is NULL. I could instead destroy `reply` and document that `reply` may be NULL when running against a non-existent collection. I could also mock a server response in `reply`, but mocking server behavior client-side sounds like trouble. 

@alcaeus I'm curious what would make more sense in PHP.

Reply:
PHP does not use the CRUD API from libmongoc but implements its own in the PHP library. As such, we're not affected by changes to this method.

Reply:
Upon taking a closer look, we'll have to consider the server reply in general: even when the collection exists, the reply will be different from what a user might expect to be returned. I'm not sure how much of this is covered by a BC promise, but users won't see the usual `count` reply document when running on 4.9+.

Reply:
As mentioned in the other thread, [CDRIVER-3933](https://jira.mongodb.org/browse/CDRIVER-3933) will be used for considering the future of the estimated document count `reply`.